test/funnels/Swapper.t.sol

```diff
@@ -0,0 +1,166 @@
+    function transfer(address, uint256) external;
+}
+
+contract CalleeMock is DssTest {
+    function swap(address src, address dst, uint256 amt, uint256, address to, bytes calldata) external {
+        GemLike(src).transfer(address(0xDEAD), amt);
+        deal(dst, address(this), amt, true);
+        GemLike(dst).transfer(to, amt);
+    }
+}
+
+contract SwapperTest is DssTest {
+    event SetLimits(address indexed gem0, address indexed gem1, uint64 hop, uint128 cap);
+    event Swap(address indexed sender, address indexed src, address indexed dst, uint256 amt, uint256 out);
+
+    AllocatorRoles public roles;
+    AllocatorBuffer public buffer;
+    Swapper public swapper;
+    UniV3SwapperCallee public uniV3Callee;
+
+    bytes32 constant ilk = "aaa";
+    bytes constant USDC_DAI_PATH = abi.encodePacked(USDC, uint24(100), DAI);
+    bytes constant DAI_USDC_PATH = abi.encodePacked(DAI, uint24(100), USDC);
+
+    address constant DAI          = 0x6B175474E89094C44Da98b954EedeAC495271d0F;
+    address constant USDC         = 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48;
+    address constant UNIV3_ROUTER = 0xE592427A0AEce92De3Edee1F18E0157C05861564;
+
+    address constant FACILITATOR = address(0x1337);
+    address constant KEEPER      = address(0xb0b);
+
+    uint8 constant SWAPPER_ROLE = uint8(1);
+
+    function setUp() public {
+        vm.createSelectFork(vm.envString("ETH_RPC_URL"));
+
+        buffer = new AllocatorBuffer();
+        roles = new AllocatorRoles();
+        swapper = new Swapper(address(roles), ilk, address(buffer));
+        uniV3Callee = new UniV3SwapperCallee(UNIV3_ROUTER);
+
+        roles.setIlkAdmin(ilk, address(this));
+        roles.setRoleAction(ilk, SWAPPER_ROLE, address(swapper), swapper.swap.selector, true);
+        roles.setUserRole(ilk, FACILITATOR, SWAPPER_ROLE, true);
+
+        swapper.setLimits(DAI, USDC, 3600 seconds, uint128(10_000 * WAD));
+        swapper.setLimits(USDC, DAI, 3600 seconds, uint128(10_000 * 10**6));
+
+        deal(DAI,  address(buffer), 1_000_000 * WAD,   true);
+        deal(USDC, address(buffer), 1_000_000 * 10**6, true);
+        buffer.approve(USDC, address(swapper), type(uint256).max);
+        buffer.approve(DAI,  address(swapper), type(uint256).max);
+    }
+
+    function testConstructor() public {
+        Swapper s = new Swapper(address(0xBEEF), "SubDAO 1", address(0xAAA));
+        assertEq(address(s.roles()),  address(0xBEEF));
+        assertEq(s.ilk(), "SubDAO 1");
+        assertEq(s.wards(address(this)), 1);
+    }
+
+    function testAuth() public {
+        checkAuth(address(swapper), "Swapper");
+    }
+
+    function testModifiers() public {
+        bytes4[] memory authedMethods = new bytes4[](1);
+        authedMethods[0] = swapper.swap.selector;
+
+        vm.startPrank(address(0xBEEF));
+        checkModifier(address(swapper), "Swapper/not-authorized", authedMethods);
+        vm.stopPrank();
+    }
+
+    function testSetLimits() public {
```

Comment:
I think we should also test that `setLimits` won't overwrite `zzz` (we could intervene the initial value).